ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,219 @@
+  , ssFee                :: !Coin
+  }  
+  
+data SnapShot c = SnapShot
+  { ssStake       :: !(Stake c)
+  , ssDelegations :: !(VMap VB VB (Credential 'Staking c) (KeyHash 'StakePool c))
+  , ssPoolParams  :: !(VMap VB VB (KeyHash 'StakePool c) (PoolParams c))
+  }  
+  
+newtype Stake c = Stake
+  { unStake :: VMap VB VP (Credential 'Staking c) (CompactForm Coin)
+  }  
+  
+newtype PoolDistr c = PoolDistr
+  { unPoolDistr ::
+      Map (KeyHash 'StakePool c) (IndividualPoolStake c)
+  }  
+  
+data IndividualPoolStake c = IndividualPoolStake
+  { individualPoolStake    :: !Rational
+  , individualPoolStakeVrf :: !(Hash c (VerKeyVRF c))
+  }  
+```
+
+> **_TODO:_** All of the maps?
+
+## Dynamics of these maps
+
+### Unified map
+
+On each application of a block (i.e. the `BBODY` rule) and each tick (i.e. the
+`TICK` rules) it can be known what entries of this map are needed to carry the
+execution. Therefore it fits the current design of UTxO-HD. 
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+  
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the unified map. Perhaps even 4 tables or a table of triplets. It
+> will have its own place on the `DbChangelog` too. The flow above describes the
+> general strategy for calling the `BBODY` rule, i.e. `(re)applyBlockOpts` and
+> for calling the `TICK` rule, i.e. `applyTickOpts`.
+
+### Stake distribution
+
+The update to this map is performed by `updateStakeDistribution`. It is known
+before calling the `BBODY` rule which UTxOs are going to be deleted and the rule
+execution logic itself knows the UTxOs that are going to be added. Therefore we
+can follow a logic similar to the above:
+
+- Before calling the ledger rule we can query for the needed entries
+- We can present the ledger with the values they asked for, but restricted to
+  the data available in the disk + changelog
+- The ledger will either provide diffs for the given values or will return
+  updated values that we can then diff with the provided ones
+- The resulting differences can be included in the current definition of the
+  `DbChangelog`.
+  
+> **_PLAN:_** Consensus will have a new table on the `LedgerTables` that will
+> represent the Incremental stake distribution (maybe two tables, one for creds
+> one for ptrs). It will have its own place on the `DbChangelog` too. The flow
+> above describes the general strategy for calling the `BBODY` rule, i.e.
+> `(re)applyBlockOpts`.
+
+### Stake snapshots
+
+The snapshots are rotated by the `snapTransition` rule. This is the most
```

Comment:
Maybe it's worth mentioning where this rule is called (I think by the `TICK` rule, right?)